rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+
+Also, note that currently, no kernel-level cache exists for block
+level functionalities, but there are plans to add it. In the meantime,
+here are a few ways to mitigate the lack of kernel-level cache:
+- hoist block-level kernel generation for it to happen once at the
+  beggining of the application
+- to handle tails/K-blocking, if generating multiple kernels is too
+  costly, one can copy A/B blocks to a temporary buffer for
+  tails/first K-block, and rely on a single block-level kernel.
+
+
+## Transforms and transpose
+
+Transormation rountines, for example to pack data in a certain layout,
+are typically hard to implement without using intrinsics or assembly.
+To facilitate packing, we will expose an out-of-place transform
+functionality.
+
+## All-in-all
+
+```c++
+
+// namespace name to be defined, leaving it general enough for additional block level APIs
+namespace dnnl {
+namespace block { 
+
+enum packing_tag {
+packed_32;
+packed_64;
+plain;
+transposed;
+}
+
+struct brgemm {
+    struct desc {
+        // Vanilla version of brgemm with no post-op or destination conversion.
+        desc(dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                float alpha, float beta);
+
+        // Advanced version with postop and datatype conversion when D type
+        // is different than C type.
+        desc(dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                data_type dtD, dim_t ldD,
+                float alpha, float beta, 
+                const brgemm_attr &attr);
+
+        // Queries for expected layouts and temporary memory
+        packing_tag get_A_tag() const;
+        packing_tag get_B_tag() const;
+        size_t get_scratchpad_size() const;
+    }
+
+    brgemm_t(const brgemm_desc_t &bd); 
+    
+    // HW context handling.
+    // This currently mimics AMX (need to clarify for SME): 
+    // - Release is static
+    // - No release necessary between different calls to set_hw_context
+    void set_hw_context() const;
+    static void release_hw_context() const;
+
+    // Execution function for the vanilla brgemm variant.
+    // we take pointers to A and B as a vector of pairs, guarantees they are the same size
+    // The batch size is the size of the vector.
+    // pointers are void*, datatypes are specified in constructor
+    // Computes C = \beta C + \alpha \sum_i A_i \cdot B_i
+    void execute(const std::vector<std::pair<void *, void *>> &A_B,
+            void *C, void *scratch = nullptr);
+
+    // Execution function for the advanced brgemm variant
+    // Here the C matrix is just an input to accumulation
+    // final result after postop/conversion will be in D
+    // Computes D = \beta C + \alpha \sum_i A_i \cdot B_i + bias
```

Comment:
so here, D has different dtype from C?

Reply:
Yes D is used for post-op and conversion fusion, hence why it has a different datatype.